python/tests/test_collections.py

```diff
@@ -0,0 +1,312 @@
+        COLLECTION_REFERENCE_KEY, COLLECTION_REFERENCE_ID
+    )
+    test_collection = collection(COLLECTION_NAME, _client)
+
+    document = test_collection.document(COLECTION_DOCUMNET_KEY)
+
+    assert isinstance(document, NumerousDocument)
+    assert document.exists is False
+
+
+def test_collection_document_returns_existing_document() -> None:
+    gql = Mock(GQLClient)
+    _client = Client(gql)
+    gql.collection_create.return_value = _collection_create_collection_reference(
+        COLLECTION_REFERENCE_KEY, COLLECTION_REFERENCE_ID
+    )
+    gql.collection_document.return_value = _collection_document_reference(
+        COLECTION_DOCUMNET_KEY
+    )
+    test_collection = collection(COLLECTION_NAME, _client)
+
+    document = test_collection.document(COLECTION_DOCUMNET_KEY)
+
+    assert isinstance(document, NumerousDocument)
+    assert document.exists
+
+
+def test_collection_document_set_data_uploads_document() -> None:
+    gql = Mock(GQLClient)
+    _client = Client(gql)
+    gql.collection_create.return_value = _collection_create_collection_reference(
+        COLLECTION_REFERENCE_KEY, COLLECTION_REFERENCE_ID
+    )
+    gql.collection_document_set.return_value = _collection_document_set_reference(
+        COLECTION_DOCUMNET_KEY
+    )
+    test_collection = collection(COLLECTION_NAME, _client)
+    document = test_collection.document(COLECTION_DOCUMNET_KEY)
+    assert isinstance(document, NumerousDocument)
+    assert document.exists is False
+
+    document.set({"test": "test"})
+
+    assert document.exists
+
+
+def test_collection_document_data_decoding_returns_dict() -> None:
+    gql = Mock(GQLClient)
+    _client = Client(gql)
+    gql.collection_create.return_value = _collection_create_collection_reference(
+        COLLECTION_REFERENCE_KEY, COLLECTION_REFERENCE_ID
+    )
+    gql.collection_document.return_value = _collection_document_reference(
+        COLECTION_DOCUMNET_KEY
+    )
+    test_collection = collection(COLLECTION_NAME, _client)
+    document = test_collection.document(COLECTION_DOCUMNET_KEY)
+
+    data = document.get()
+
+    assert isinstance(document, NumerousDocument)
+    assert document.exists
+    assert data
+    if data is not None:
+        assert data["test"] == "test"
+
+
+def test_collection_document_delete_marks_document_exists_false() -> None:
+    gql = Mock(GQLClient)
+    _client = Client(gql)
+    gql.collection_create.return_value = _collection_create_collection_reference(
+        COLLECTION_REFERENCE_KEY, COLLECTION_REFERENCE_ID
+    )
+    gql.collection_document.return_value = _collection_document_reference(
+        COLECTION_DOCUMNET_KEY
+    )
+    test_collection = collection(COLLECTION_NAME, _client)
+    document = test_collection.document(COLECTION_DOCUMNET_KEY)
+    if document.document_id is not None:
+        gql.collection_document_delete.return_value = _collection_document_delete_found(
+            document.document_id
+        )
```

Comment:
In general conditional logic in a test is a code smell. Why would this test run differently sometimes?